doc/proposals/scorecard-custom-tests-2.md

```diff
@@ -0,0 +1,293 @@
+
+### Scorecard Configuration
+
+Note that it would also be possible to add a resource limit to the scorecard configuration as shown in the example below.  Those resource limits would be applied to custom tests that are executed allowing for some control as to how much resources are available for the test execution.
+
+Custom test Pod manifests would be accessible from a local disk location to scorecard.  Scorecard would obtain the custom test manifests as part of its bundle image processing since the bundle image is passed into scorecard as a configuration flag.  
+
+Scorecard would assume that the scorecard configuration file and test
+manifests would be located in the bundle image at the following location:
+```
+/bundle/scorecard/.osdk-scorecard.yaml
+/bundle/scorecard/tests
+```
+
+Test Pod manifests would be configured from within the scorecard configuration file.  The following is a scorecard configuration file example using this scheme:
+
+```
+scorecard:
+  # Setting a global scorecard option
+  output: text
+  tests:
+    init-timeout: 60
+    resources:
+      limits:
+        cpu: "1"
+        memory: "200Mi"
+      requests:
+        cpu: 750m
+        memory: "100Mi"
+      test-path:
+        - "/bundle/scorecard/tests"
+  …
+```
+
+
+### Custom Test Execution
+
+Scorecard tests would be executed from the SDK scorecard CLI as is the case today.  Scorecard would execute the tests, capture the results, and provide feedback to the end-user by means of the current scorecard API (e.g. ScorecardTestResult).  Command line flags would allow the end-user a means to specify which tests are executed and upon which test environment to run the tests.  For example:
+```
+operator-sdk scorecard -o text --selector=suite=isvcustom --bundle-image='some/bundle/image:v0.0.1'
+```
+
+Running the above command would have scorecard perform the following:
+ 
+ * fetch the test configuration and test manifests from the bundle image
+ * select the tests to run based on the test manifest labels and the selector
+ * iterate through each CR within the bundle image CSV
+ * execute the test image Pod for each selected test passing in the current CR as an environment variable
+ * fetch the test image Pods log output for aggregation of test results
+ * display the test results 
+
+### Custom Test Environments
+
+Scorecard would allow you to run tests on a local Kube cluster as is the case today.  Provisioning the test cluster is external to the scope of the scorecard itself.  
+
+### Custom Test Labeling
+
+As is the case today, custom tests would allow end users to specify custom labels to each test so that the scorecard could select which tests are run based on a labeling scheme.  Here is an example of user added labels to a custom test step:
```

Comment:
I'm going to make some statements that will probably result in more discussion and thought. Tell me what you think.

1. Tests are defined as Pods
2. Each pod that runs can return multiple test results.
3. Scorecard test results have names, descriptions, and labels.
4. Pod `metadata.labels` correspond to the scorecard test labels for the purposes of selection. So labels apply at the pod level to all of the tests returned by the image when executed.
5. Scorecard uses the pod `name` when displaying tests selected by the labels with `--list`
6. When the pod runs, it returns multiple results, which will each have different names defined by the test image. 
7. Scorecard uses something in the pod (maybe an annotation?) to display test descriptions selected by labels with `--list`

I'm getting a little hung up on the issue of having two layers where names, labels, and descriptions are relevant. I think we need more specifics around this.

Would it cause too much cluster and configuration churn if we said that a pod can return only one test result?

Reply:
yeah, the labeling breaks down if we support multiple tests in a single pod spec, to make it work you would need to implement a set-based label expression "test in (test1,test2)", but then you would also need to implement a richer configuration to have multiple test descriptions and test lables, it turns into a mess.

So yes, I think we just go with 1 pod = 1 test to simplify for both the scorecard and for the end user as well.  So that would be my vote.  

The test description I think is just another metadata.label if we go with a single test per Pod approach.

Reply:
I think label values have a max length of 63 characters (unless that's been changed recently?). If that's the case, that's why I suggested an annotation for description.

Reply:
good point.